File: src/main/java/model/HealthInfo.java
package model;

@Model("health_info")
public class HealthInfo {
    public HealthInfo(Person person, String tel, boolean danger14, boolean abroad14, boolean touchIllness14, boolean ill, Integer status) {
        this.person = person;
        this.tel = tel;
        this.danger14 = danger14;
        this.abroad14 = abroad14;
        this.touchIllness14 = touchIllness14;
        this.ill = ill;
        this.status = status;
    }

    @Field("uid")
    private Person person;

    @Field("tel")
    private String tel;

    @Field("if_danger_14")
    private boolean danger14;

    @Field("if_abroad_14")
    private boolean abroad14;

    @Field("if_touch_illness")
    private boolean touchIllness14;

    @Field("if_ill")
    private boolean ill;

    @Field("status")
    private Integer status;

    @Field("color")
    private String color;

    public HealthInfo() {
    }

    public CodeColor getCodeColor() {
        var codeColor = CodeColor.GREEN;
        var abnormalCount = getStatus().getAbnormalCount();

        if (danger14 || abroad14 || abnormalCount == 1) {
            codeColor = CodeColor.YELLOW;
        }

        if (touchIllness14 || ill || abnormalCount >= 2) {
            codeColor = CodeColor.RED;
        }

        return codeColor;
    }

    public String getTel() {
        return tel;
    }

    public void setTel(String tel) {
        this.tel = tel;
    }

    public boolean isDanger14() {
        return danger14;
    }

    public void setDanger14(boolean danger14) {
        this.danger14 = danger14;
    }

    public boolean isAbroad14() {
        return abroad14;
    }

    public void setAbroad14(boolean abroad14) {
        this.abroad14 = abroad14;
    }

    public boolean isTouchIllness14() {
        return touchIllness14;
    }

    public void setTouchIllness14(boolean touchIllness14) {
        this.touchIllness14 = touchIllness14;
    }

    public boolean isIll() {
        return ill;
    }

    public void setIll(boolean ill) {
        this.ill = ill;
    }

    public Status getStatus() {
        return new Status(status);
    }

    public void setStatus(Integer status) {
        this.status = status;
    }

    public Person getPerson() {
        return person;
    }

    public void setPerson(Person person) {
        this.person = person;
    }

    public String getColor() {
        return color;
    }

    public void setColor(String color) {
        this.color = color;
    }

    @Override
    public String toString() {
        return "HealthInfo{" +
                "person=" + person +
                ", tel='" + tel + '\'' +
                ", danger14=" + danger14 +
                ", abroad14=" + abroad14 +
                ", touchIllness14=" + touchIllness14 +
                ", ill=" + ill +
                ", status=" + new Status(status) +
                ", color='" + color + '\'' +
                '}';
    }
}

File: src/main/java/model/HealthInfo.kt
package model

import java.text.MessageFormat

@Model("health_info")
class HealthInfo {
    constructor(person: Person, tel: String, danger14: Boolean, abroad14: Boolean, touchIllness14: Boolean, ill: Boolean, status: Int) {
        this.person = person.personId
        this.tel = tel
        isDanger14 = danger14
        isAbroad14 = abroad14
        isTouchIllness14 = touchIllness14
        isIll = ill
        this.status = status
    }

    @Field("uid")
    var person: String = ""

    @Field("tel")
    var tel: String = ""

    @Field("if_danger_14")
    var isDanger14 = false

    @Field("if_abroad_14")
    var isAbroad14 = false

    @Field("if_touch_illness")
    var isTouchIllness14 = false

    @Field("if_ill")
    var isIll = false

    @Field("status")
    private var status: Int = 0

    @Field("color")
    var color: String = CodeColor.RED.toString()

    constructor() {}

    val codeColor: CodeColor
        get() {
            var codeColor = CodeColor.GREEN
            val abnormalCount = getStatus().abnormalCount
            if (isDanger14 || isAbroad14 || abnormalCount == 1) {
                codeColor = CodeColor.YELLOW
            }
            if (isTouchIllness14 || isIll || abnormalCount >= 2) {
                codeColor = CodeColor.RED
            }
            return codeColor
        }

    fun getStatus() = Status(status)

    fun getPerson() = SQL().query(Person::class.java, "uid" to person)

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is HealthInfo) return false

        if (person != other.person) return false

        return true
    }

    override fun hashCode(): Int {
        return person.hashCode()
    }

    override fun toString(): String {
        return "HealthInfo(person='$person', tel='$tel', isDanger14=$isDanger14, isAbroad14=$isAbroad14, isTouchIllness14=$isTouchIllness14, isIll=$isIll, status=$status, color='$color')"
    }

}